src/FilterInterface.php

```diff
@@ -4,14 +4,26 @@
 
 namespace Laminas\Filter;
 
+/** @template TFilteredValue */
 interface FilterInterface
 {
     /**
      * Returns the result of filtering $value
      *
+     * @template T
+     * @param T $value
+     * @return TFilteredValue|T
```

Comment:
Shouldn't this be `TFilteredValue&T`?

We're trying to restrict the type, not expand it :thinking: 

This can be verified wtih `$v = (new AllowList(['list' => ['a', 'b', 'c']))->filter(123)`, which becomes `123|'a'|'b'|'c'|null` here, while we want `'a'|'b'|'c'|null` to be the actual returned type.

I believe we need to check this a bit better :mag:

Reply:
Is every filter supposed to return `|null` BTW? Unsure about that detail...

Reply:
This filter is a bit odd. Typically, filter failure returns the un-filtered value, so `$stringToLower->filter($someArray): array`. So, StringToLower for example would be `@implements FilterInterface<string>` and return `string|T` - this AllowList filter is odd because in a success scenario, it returns T vs null in a failure scenario, hence the reversal of the failure/success type.

The AllowList filter specifically could be improved with a template for the list perhaps something like:

```php
/**
 * @template MyList of list<mixed>
 * @implements FilterInterface<value-of<MyList>>
 */
final class AllowList implements FilterInterface
{
  /** @param MyList $list */
  public function __construct(private array $list) {}
}
```

And I guess we'd then need to override the filter method docs to return `TFilteredValue|null`. That would be more precise right?

I assumed that the current docs as they are here would have yielded:
```php
$v = (new AllowList(['list' => ['a', 'b', 'c']))->filter('a'); // 'a'|null
$v = (new AllowList(['list' => ['a', 'b', 'c']))->filter('z'); // 'z'|null
```

Reply:
Actually, I don't think it's possible to restrict the return type to members of the list, and it does currently restrict the type to `T|null` which accurately describes the behaviour of the filter.

Adding a template to describe the list is the only way I know of to get `@return value-of<TList>` to work, i.e. `@return value-of<$this->property>` doesn't work, _(or I'm wrong 🤷‍♂️)_ - If I use a template, you'd have to annotate every usage of the filter with `@var AllowList<list{'a'|'b'|'c'}>` right?

I might need a bit of help here to refine the filter to return anything beyond `T|null`

Reply:
Is `value-of<TList>` not an actual template type?

I need to think of it myself too btw, just thinking that current union type is a bit too wide

Reply:
@gsteel does this example of yours work BTW?

```php
/**
 * @template MyList of list<mixed>
 * @implements FilterInterface<value-of<MyList>>
 */
final class AllowList implements FilterInterface
{
  /** @param MyList $list */
  public function __construct(private array $list) {}
}
```

I think this is very close to what we want, and verifiable via static analysis test code

Reply:
No - it doesn't work… I tried a bunch of things but couldn't figure out how to get inference from `value-of<$this->list>` _(value-of<$property> is not supported afaict)_, also, without annotating all of the filters you _use_ to satisfy `TList`, you end up with mixed anyhow, and a bunch of psalm issues for missing templates. It's not like we can put the template on the constructor either, because then inference is lost in the `filter` method which is where we want it - i.e. we can't do constructor property promotion because the convention of a single `$options` argument.

TBH, the other filters should be a lot more straightforward - I started with this one because it's the first one in the list alphabetically